### PR TITLE
Uptake new shape of ReportStreamError

### LINKED ISSUE #3081: Correctly identify RS warnings and store them in test-event-publishing-exceptions
### Description
Report Stream changed their response object, which resulted in our `QueueBatchedReportStreamUploader ` azure function to incorrectly put messages with no warning body on the `test-event-publishing-exceptions` queue
example: `{"isError":false}`
this causes the `ReportStreamExceptionHandler` azure function to experience issues when uploading the warnings to SR report stream callback endpoint.

### Proposed solution 
don't put responses without warning onto the `test-event-publishing-exceptions` queue
### Additional context
[Slack convo](https://skylight-hq.slack.com/archives/C01LTSNKEPP/p1637697374280700)
reach out to @zdeveloper for more context

### PR DESCRIPTION
**Pull request by mehansen**
# DEVOPS PULL REQUEST

## Related Issue

- Resolves #3081 

## Changes Proposed

Update the way we report exceptions included in the ReportStream response to when we upload test results. The response shape changed so we need to uptake this new shape.
Also added some logic to stop upload if there are no valid messages to include, because otherwise the upload will fail.

## Additional Information

- `trackingIds` in the report stream response `error` and `warning` objects correspond to the Result_ID field in the message we send them, which is the test result internal ID, so we can map any error with `trackingIds` to the right test event
-  errors at the report level (`scope: report`) do not have `trackingIds`, so for now we just log it in Azure and do not send it to the publishing-exceptions queue to be persisted in our table
    - judging from Metabase we have never persisted report-scope exceptions in our table so this isn't a regression, but it might be nice to explore a way to handle this in the future, either by adding a row for every test result in the request that generated that exception or making some db/backend changes to allow us to link an exception to a group of test events

## Testing

- Drop an incorrectly formatted message in the publishing queue, wait <= 2 minutes for QueueBatchedReportStreamUploader to fire, then check the publishing-exceptions queue and verify the right exception message has been added to the queue 
- Did this "over-the-shoulder" with @nathancrtr in dev
 
### Infrastructure
- [ ] Consult the results of the `terraform-plan` job inside the "Terraform Checks" workflow run for this PR. Confirm that there are no unexpected changes!

### Security
- [n/a] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [ ] Any dependencies introduced have been vetted and discussed

## Cloud
- [ ] Oncall has been notified if this change is going in after-hours
- [ ] If there are changes that cannot be tested locally, this has been deployed to our Azure `test`, `dev`, or `pentest` environment for verification

### CODE REVIEW

File: ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/lib.ts [hunk 1]
```diff
@@ -173,22 +174,26 @@ export async function reportExceptions(
...
-      details,
-    }))
-    .concat(
-      response.errors.map(({ id, details }) => ({
-        testEventInternalId: id,
-        isError: true,
-        details,
-      }))
-    );
+  const payloads: SimpleReportReportStreamResponse[] = response.warnings.flatMap(w => responsesFrom(w, false))
```
Comment (nathancrtr) [inline ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/lib.ts:177]: I forgot that `Array.prototype.flatMap` finally landed in ES 😁 very nice

**Review by DanielSass — APPROVED** [review]
looks good!

File: ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/index.ts [hunk 2]
```diff
@@ -53,6 +53,11 @@ const QueueBatchedTestEventPublisher: AzureFunction = async function (
     });
   }
 
+  if (parseSuccessCount < 1) {
+    context.log(`Successfully parsed message count of ${parseSuccessCount} is less than 1; aborting`);
+    return;
+  }
```
Comment (nathancrtr) [inline ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/index.ts:59]: What happens to messages that are dequeued but then fail to parse? Will they go back on the queue or just disappear?

File: ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/lib.ts [hunk 3]
```diff
@@ -173,22 +174,24 @@ export async function reportExceptions(
...
     payloads.map((p) =>
       queueClient.sendMessage(Buffer.from(JSON.stringify(p)).toString("base64"))
     )
   );
 }
+
+const responsesFrom = function(context: Context, err: ReportStreamError, isError: boolean):SimpleReportReportStreamResponse[] {
+  if (err.trackingIds) {
+    return err.trackingIds.map(id => ({
+      testEventInternalId: id,
```
Comment (nathancrtr) [inline ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/lib.ts:189]: Is there anything else we or ReportStream need to do to make sure the `trackingElement` is the internal ID of the `TestEvent`?

Comment (nathancrtr) [reply ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/lib.ts:189]: Nvm, I remember we looked at this 👍

File: ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/index.ts (hunk 2 above)
Comment (mehansen) [reply ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/index.ts:59]: when we dequeue messages, we're not actually deleting them, seems like. that doesn't happen until we call `deleteSuccessfullyParsedMessages`, on the condition that our upload is successful.

additionally, within that method we also skip deleting messages that failed to parse
https://github.com/CDCgov/prime-simplereport/blob/f59fe3971a88d95cac960f8f3bc08dd54763e444/ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/lib.ts#L144-L149

so essentially they stay on the queue and we keep trying to parse them until they expire 7 days after initial insertion. [after 5 dequeues a message gets logged to the poison queue](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-storage-queue-trigger?tabs=in-process%2Cextensionv5&pivots=programming-language-javascript#poison-messages), but it doesn't look like we do anything with this info. maybe we could have a ticket to alert us if something gets added to the poison queue

Comment (mehansen) [reply ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/index.ts:59]: ignore the last part, that only applies to functions triggered by queues, like our exception handler function

Comment (nathancrtr) [reply ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/index.ts:59]: Thanks for the explanation 🙌 that makes sense. I actually didn't know about the "poison" queue but that's exactly what I was hoping for

**Pull request merged**